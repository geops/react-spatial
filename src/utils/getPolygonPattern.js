import { DEVICE_PIXEL_RATIO } from 'ol/has';

const getPolygonPattern = (patternId = 1, color = [235, 0, 0, 1]) => {
  if (patternId === 1) {
    return color;
  }

  const canvasElement = document.createElement('canvas');
  const pixelRatio = DEVICE_PIXEL_RATIO;

  canvasElement.width = 20 * pixelRatio;
  canvasElement.height = 20 * pixelRatio;

  let pattern = {};
  const ctx = canvasElement.getContext('2d');
  ctx.strokeStyle = `rgba(${color.toString()})`;
  ctx.fillStyle = `rgba(${color.toString()})`;
  ctx.lineWidth = 3;

  switch (patternId) {
    case 2:
      /* Hatched pattern */
      /* Ascending line */
      ctx.beginPath();
      ctx.moveTo(0, canvasElement.height);
      ctx.lineTo(0, canvasElement.height - ctx.lineWidth / 2);
      ctx.lineTo(canvasElement.width - ctx.lineWidth / 2, 0);
      ctx.lineTo(canvasElement.width, 0);
      ctx.lineTo(canvasElement.width, ctx.lineWidth / 2);
      ctx.lineTo(ctx.lineWidth / 2, canvasElement.height);
      ctx.lineTo(ctx.lineWidth / 2, canvasElement.height);
      ctx.fill();
      ctx.closePath();

      /* Descending line */
      ctx.beginPath();
      ctx.moveTo(0, 0);
      ctx.lineTo(0, ctx.lineWidth / 2);
      ctx.lineTo(canvasElement.width - ctx.lineWidth / 2, canvasElement.height);
      ctx.lineTo(canvasElement.width, canvasElement.height);
      ctx.lineTo(canvasElement.width, canvasElement.height - ctx.lineWidth / 2);
      ctx.lineTo(ctx.lineWidth / 2, 0);
      ctx.lineTo(0, 0);
      ctx.fill();
      ctx.closePath();

      pattern = ctx.createPattern(canvasElement, 'repeat');
      pattern.canvas = canvasElement;
      break;
    case 3:
      /* Shade ascending pattern */
      /* Corner triangle */
      ctx.beginPath();
      ctx.moveTo(0, 0);
      ctx.lineTo(0, ctx.lineWidth / 2);
      ctx.lineTo(ctx.lineWidth / 2, 0);
      ctx.fill();
      ctx.closePath();

      /* Ascending line */
      ctx.beginPath();
      ctx.moveTo(0, canvasElement.height);
      ctx.lineTo(0, canvasElement.height - ctx.lineWidth / 2);
      ctx.lineTo(canvasElement.width - ctx.lineWidth / 2, 0);
      ctx.lineTo(canvasElement.width, 0);
      ctx.lineTo(canvasElement.width, ctx.lineWidth / 2);
      ctx.lineTo(ctx.lineWidth / 2, canvasElement.height);
      ctx.lineTo(ctx.lineWidth / 2, canvasElement.height);
      ctx.fill();
      ctx.closePath();

      /* Corner triangle */
      ctx.beginPath();
      ctx.moveTo(canvasElement.width, canvasElement.height);
      ctx.lineTo(canvasElement.width, canvasElement.height - ctx.lineWidth / 2);
      ctx.lineTo(canvasElement.width - ctx.lineWidth / 2, canvasElement.height);
      ctx.fill();
      ctx.closePath();

      pattern = ctx.createPattern(canvasElement, 'repeat');
      pattern.canvas = canvasElement;
      break;
    case 4:
      /* Shade descending pattern */
      /* Corner triangle */
      ctx.beginPath();
      ctx.moveTo(canvasElement.width, 0);
      ctx.lineTo(canvasElement.width, ctx.lineWidth / 2);
      ctx.lineTo(canvasElement.width - ctx.lineWidth / 2, 0);
      ctx.fill();
      ctx.closePath();

      /* Descending line */
      ctx.beginPath();
      ctx.moveTo(0, 0);
      ctx.lineTo(0, ctx.lineWidth / 2);
      ctx.lineTo(canvasElement.width - ctx.lineWidth / 2, canvasElement.height);
      ctx.lineTo(canvasElement.width, canvasElement.height);
      ctx.lineTo(canvasElement.width, canvasElement.height - ctx.lineWidth / 2);
      ctx.lineTo(ctx.lineWidth / 2, 0);
      ctx.lineTo(0, 0);
      ctx.fill();
      ctx.closePath();

      /* Corner triangle */
      ctx.beginPath();
      ctx.moveTo(0, canvasElement.height);
      ctx.lineTo(0, canvasElement.height - ctx.lineWidth / 2);
      ctx.lineTo(ctx.lineWidth / 2, canvasElement.height);
      ctx.fill();
      ctx.closePath();

      pattern = ctx.createPattern(canvasElement, 'repeat');
      pattern.canvas = canvasElement;
      break;
    default:
  }

  if (patternId === 0) {
    pattern.empty = true;
  }

  pattern.id = patternId;
  pattern.color = color;

  return pattern;
};

export default getPolygonPattern;
