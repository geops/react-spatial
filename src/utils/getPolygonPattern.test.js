import getPolygonPattern from './getPolygonPattern';

describe('getPolygonPattern()', () => {
  test('render pattern with default properties (id=1, color = [235, 0, 0, 1])', () => {
    const color = [235, 0, 0, 1];
    const pattern = getPolygonPattern();
    expect(pattern).toEqual(color);
    expect(pattern.id).toBe();
    expect(pattern.color).toBe();
    expect(pattern.empty).toBe();
    expect(pattern.canvas).toBe();
  });

  test('render pattern 0 (no fill) color and (light blue) opacity', () => {
    const id = 0;
    const color = [0, 60, 80, 0.41000000000000003];
    const pattern = getPolygonPattern(id, [0, 60, 80, 0.41000000000000003]);
    expect(pattern.id).toBe(id);
    expect(pattern.color).toEqual(color);
    expect(pattern.empty).toBe(true);
    expect(pattern.canvas).toBe();
  });

  test('render pattern 1 (full by color) color and (light blue) opacity', () => {
    const id = 1;
    const color = [0, 60, 80, 0.41000000000000003];
    const pattern = getPolygonPattern(id, [0, 60, 80, 0.41000000000000003]);
    expect(pattern).toEqual(color);
    expect(pattern.id).toBe();
    expect(pattern.color).toBe();
    expect(pattern.empty).toBe();
    expect(pattern.canvas).toBe();
  });

  test('render pattern 2 (cross) color and (light blue) opacity', () => {
    const id = 2;
    const color = [0, 60, 80, 0.41000000000000003];
    const pattern = getPolygonPattern(id, [0, 60, 80, 0.41000000000000003]);
    expect(pattern.id).toBe(id);
    expect(pattern.color).toEqual(color);
    expect(pattern.empty).toBe();
    expect(pattern.canvas).toMatchSnapshot();
  });

  test('render pattern 3 (diagonal line from bottom-left tot top-right) with color (light blue) and opacity', () => {
    const id = 3;
    const color = [0, 60, 80, 0.41000000000000003];
    const pattern = getPolygonPattern(id, [0, 60, 80, 0.41000000000000003]);
    expect(pattern.id).toBe(id);
    expect(pattern.color).toEqual(color);
    expect(pattern.empty).toBe();
    expect(pattern.canvas).toMatchSnapshot();
  });

  test('render pattern 4 (diagonal line from top-left to bottom-right) with color (light blue) and opacity', () => {
    const id = 4;
    const color = [0, 60, 80, 0.41000000000000003];
    const pattern = getPolygonPattern(id, [0, 60, 80, 0.41000000000000003]);
    expect(pattern.id).toBe(id);
    expect(pattern.color).toEqual(color);
    expect(pattern.empty).toBe();
    expect(pattern.canvas).toMatchSnapshot();
  });
});
