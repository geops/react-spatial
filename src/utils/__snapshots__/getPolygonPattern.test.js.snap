// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getPolygonPattern() render pattern 2 (cross) color and (light blue) opacity 1`] = `
<canvas
  data-snapshot-image="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABQAAAAUCAYAAACNiR0NAAAABmJLR0QA/wD/AP+gvaeTAAABKElEQVQ4jZ3UQUrDUBDG8f/kGUohyQEUpS2CrtwKZuPWRdNV7+DpCkbEExRxL64Eg+AFmm7apHkuNJIa00w6yzD58b6XyQhhdItlySaLeX5csE9dTz3ydYS1gQOkCB4H7pgw8jtjYeSTrSZYGwCpg9u7Q2QB+FgmXN4EnU4mMv55d0me3RuS1zWji3dsMQB8xAw5HCV8vq20MatXZgA6ow0YgPlt0qI7sG1Qg7ZgAKKO1PeKNqwZ/IsiORQWxG2bWfPfw+34nGO5AnsC9oNNPtv1AziNYK3LUbV3iCwWij0j1z/KjF7/RTOndbBpNJRzalRYWQrUqDElajphClQ6Yy1XZDg6nQIB34s25ukhVWHlSY/PEmCI4OOYgUNlOTKP9VhZ8zitLukvrS9K3yr02NEAAAAASUVORK5CYII="
  height="20"
  width="20"
/>
`;

exports[`getPolygonPattern() render pattern 3 (diagonal line from bottom-left tot top-right) with color (light blue) and opacity 1`] = `
<canvas
  data-snapshot-image="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABQAAAAUCAYAAACNiR0NAAAABmJLR0QA/wD/AP+gvaeTAAAApUlEQVQ4ja3UvQ3CMBiE4deSS8xA0LAApqJjSUzhGbJOQhNMFIpEioD82J99Azy6a05xsDe64Kh8jTSn6473y9L3e43iTuUbMXa0htBawACNEkM/zeh50gUnB2cwKl/LwAUMIB1cwdLBDSwNjMDiwUgsDkzAtsFEbB0UYMugEJsHM7B/MBP7BgtgE1gIG8CCGIAmtBfGc0Thcs9WA2Zs9sjChqXnDztL4Jv7WOXKAAAAAElFTkSuQmCC"
  height="20"
  width="20"
/>
`;

exports[`getPolygonPattern() render pattern 4 (diagonal line from top-left to bottom-right) with color (light blue) and opacity 1`] = `
<canvas
  data-snapshot-image="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABQAAAAUCAYAAACNiR0NAAAABmJLR0QA/wD/AP+gvaeTAAAA30lEQVQ4ja3UPQ4BURSG4ffOIRp3FiCRMAqdVjGNVmFU9mBvKhRajcIGJBKJScQG0Bg/V4NE/M3cma865xRPTvUp/KCP4cDlNGI+2WGbVq/IOQocYI+iSC7fwQ+0NTgdHDBmLNQaa8y1ArgYPEpeyHZ1tEI3y0gIFxHeE9UoqaZBBSBLVJ5TRqi8bBmg8nZJib6DKdHPYAr0O2iJ/gYt0P9gQjQemACND8ZEk4Ex0OTgH1RZgY/cSxVj3EdJC36g2Swj60/L9RCootA4UnEwdGm2XesvZ6M9+cIQpXaAvgH5/sBe6ItkwwAAAABJRU5ErkJggg=="
  height="20"
  width="20"
/>
`;
